collector-api/internal/api/prometheus.go

```diff
@@ -43,35 +48,129 @@ type prometheusClient struct {
 	endpoint url.URL
 }
 
+type promResult struct {
+	Metric    map[string]string
+	Value     float64
+	Timestamp time.Time
+}
+
 func (c *prometheusClient) Do(req *http.Request) (*http.Response, error) {
 	req.Header.Set("User-Agent", "autodba")
 	return c.Client.Do(req)
 }
 
+func (c *prometheusClient) QueryWithRetry(query string, ts time.Time, maxRetries int) ([]promResult, error) {
+	result, err := c.Query(query, ts)
+	if err == nil {
+		return result, nil
+	}
+
+	log.Printf("Initial Prometheus query failed: %v", err)
+
+	if !strings.Contains(err.Error(), "503") {
+		return nil, fmt.Errorf("query Prometheus failed: %w", err)
+	}
+
+	log.Printf("Received 503 error, starting retry attempts")
+
+	for retries := 0; retries < maxRetries; retries++ {
+		waitTime := time.Second * time.Duration(retries+1)
+		log.Printf("Retry attempt %d after waiting %v", retries+1, waitTime)
+
+		time.Sleep(waitTime)
+		result, err = c.Query(query, ts)
+
+		if err == nil {
+			log.Printf("Retry attempt %d successful", retries+1)
+			return result, nil
+		}
+		log.Printf("Retry attempt %d failed: %v", retries+1, err)
+	}
+
+	return nil, fmt.Errorf("query Prometheus failed after %d retries: %w", maxRetries, err)
+}
+
+func (c *prometheusClient) Query(query string, t time.Time) ([]promResult, error) {
+	// Create v1 API client
+	client, err := api.NewClient(api.Config{
+		Address: fmt.Sprintf("%s://%s", c.endpoint.Scheme, c.endpoint.Host),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error creating API client: %w", err)
+	}
+	v1api := v1.NewAPI(client)
+
+	// Execute query using v1 API
+	ctx := context.Background()
+	result, warnings, err := v1api.Query(ctx, query, t)
+	if err != nil {
+		return nil, fmt.Errorf("execute query: %w", err)
+	}
+
+	if len(warnings) > 0 {
+		log.Printf("Warnings from query (\"%s\"): %v", query, warnings)
+	}
+
+	// Convert result to promResult format
+	var results []promResult
+	if vector, ok := result.(model.Vector); ok {
```

Comment:
will this only ever handle queries that return vector results ?

Reply:
Up until now, it's only `Vector`, but you're right that we should handle more types, as it's a generic function. Adde that functionality in [3333b9d](https://github.com/crystaldb/autodba/pull/71/commits/3333b9de47008a89f20be55ec9a395debc923619)